json_serializable/lib/src/json_key_with_conversion.dart

```diff
@@ -157,6 +158,25 @@ class JsonKeyWithConversion extends JsonKey {
     assert(!this.includeIfNull || !this.disallowNullValue);
   }
 
+  static String _processName(JsonSerializable classAnnotation,
+      String jsonKeyNameValue, FieldElement fieldElement) {
+    if (jsonKeyNameValue != null) {
+      return jsonKeyNameValue;
+    }
+
+    switch (classAnnotation.fieldRename) {
```

Comment:
here's another possible way to handle this which leaves more room for extension:

change the `fieldRename` argument from an `enum` to a `String Function(String)`. Change the top-level method `snakeCase` to a static method `FieldRename.snake`.

- Usage looks identical: `@JsonSerializable(fieldRename: FieldRename.snake)`
- No need for a `switch` or needing to touch lots of points of code to add new options.
- Users can add their own rename logic by implementing a `String Function(String)`.

Reply:
@natebosch , you're crazy – like a 🦊 

Would likely help headed toward https://github.com/dart-lang/json_serializable/issues/280 , too

Reply:
Well...I'll try. The function is defined in user code, but needs to be invoked in the context of the build action – so...maybe we can make it work with mirrors?

We'll see...

Reply:
hmm, yeah that might not be feasible - it puts us in really weird places when it comes to invalidation of the build script - we might not invalidate the outputs if the users changes the implementation of the rename function.

We could only do this, I think, if the rename were lazy and done at serialization time - which also means we'd need a pair of functions to go in both directions...